Client/Experiments/Messaging/GleanPlumbMessageStore.swift

```diff
@@ -0,0 +1,129 @@
+    /// Do the bookkeeping for message dismissed Counts and expiry.
+    func onMessageDismissed(_ message: GleanPlumbMessage)
+
+    /// Handle all points of expiry and Telemetry, and returns an updated metadata object for use.
+    func onMessageExpired(_ message: GleanPlumbMessageMetaData, shouldReport: Bool) -> GleanPlumbMessageMetaData
+
+}
+
+class GleanPlumbMessageStore: GleanPlumbMessagingStoreProtocol {
+
+    // MARK: - Properties
+
+    private let decoder = JSONDecoder()
+    private let encoder = JSONEncoder()
+
+    // MARK: - MessageStoreProtocol methods
+
+    /// Returns the metadata that persists on system. If there's none, it returns default data.
+    func getMessageMetadata(messageId: String) -> GleanPlumbMessageMetaData {
+
+        /// Return preexisting Message Metadata.
+        if let metadata = get(key: messageId) {
+            return metadata
+        }
+
+        return GleanPlumbMessageMetaData(Id: messageId,
+                                         impressions: 0,
+                                         dismissals: 0,
+                                         isExpired: false)
+    }
+
+    /// Update message metadata and persist that information.
+    func onMessageDisplayed(_ message: GleanPlumbMessage) {
+        var messageToTrack = message.metadata
+        let messageImpressions = message.metadata.impressions
+        let messageMaxCount = message.style.maxDisplayCount
+
+        messageToTrack.impressions += 1
+
+        /// Determine if it's expired.
+        if messageImpressions >= messageMaxCount || messageToTrack.isExpired {
+            messageToTrack = onMessageExpired(messageToTrack, shouldReport: true)
+        }
+
+        set(key: message.Id, metadata: messageToTrack)
+    }
+
+    /// For the MVP, we always expire the message.
+    func onMessagePressed(_ message: GleanPlumbMessage) {
+        let messageToTrack = onMessageExpired(message.metadata, shouldReport: false)
+
+        set(key: message.Id, metadata: messageToTrack)
+    }
+
+    /// Depending on the surface, we may do different things with dismissal. But for the MVP,
+    /// dismissal expires the message.
+    func onMessageDismissed(_ message: GleanPlumbMessage) {
+        var messageToTrack = onMessageExpired(message.metadata, shouldReport: false)
+
+        messageToTrack.dismissals += 1
+
+        set(key: message.Id, metadata: messageToTrack)
+    }
+
+    /// Updates a message's metadata and reports expiration Telemetry when applicable.
+    /// A message expires in three ways (dismissal, interaction and max impressions), but only
+    /// impressions should report an expired message.
+    func onMessageExpired(_ message: GleanPlumbMessageMetaData, shouldReport: Bool) -> GleanPlumbMessageMetaData {
+        var messageToTrack = message
+        messageToTrack.isExpired = true
+
+        if shouldReport {
+            TelemetryWrapper.recordEvent(category: .information,
+                                         method: .view,
```

Comment:
Is there a reason you used method `.view` here but on malformed you used `.application`?